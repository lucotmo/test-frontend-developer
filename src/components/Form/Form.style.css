.form{
  display: flex;
  flex-flow: column wrap;
  justify-content: center;
  width: 100%;
  margin-top: 3rem;
  padding: 1rem;
  border-radius: 4px;
  background-color: white;
  box-shadow: 1px 1px 3px 2px rgba(238, 238, 238, 1);
}

.form__title{
  text-align: center;
  margin: .5rem 0;
  color: #333;
}

.form__welcome{
  margin: .5rem 0;
  color: #777;
}

.form__welcome strong{
  color: #333;
}

.form__errors{
  margin: .5rem 0;
  padding: .5rem;
  color: white;
  text-align: center;
  border-radius: 4px;
  background-color: #d71209;
}