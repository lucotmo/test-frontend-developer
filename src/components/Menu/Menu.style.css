.container{
  margin: 0;
  padding: 1rem;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  list-style: none;
  background-color: #333;
}

.menu{
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  margin: 0;
  padding: 1rem;
  display: flex;
  flex-flow: column wrap;
  justify-content: space-between;
  transform: translateY(-120%);
  list-style: none;
  z-index: 10;
  transition: .5s all;
  background-color: #333;
}

.menu__link{
  margin: 0.1rem 0 0.1rem 0;
  padding: .5rem;
  border-radius: 4px;
  cursor: pointer;
  background-color: white;
}

.activelink {
  font-weight: 700;
  letter-spacing: 0.05rem;
  color: white;
  background-color: #d71209;
}

.menu__button{
  width: 32px;
  height: 28px;
  padding: 4px;
  right: 1rem;
  display: flex;
  flex-flow: column wrap;
  justify-content: space-between;
  align-items: center;
  border-radius: 4px;
  cursor: pointer;
  z-index: 20;
  background-color: #333;
}

.menu__line{
  width: 100%;
  height: 5px;
  background-color: white;
}

.menu--show{
  transform: translateY(40%);
}

@media (min-width: 920px) {
  .container{
    flex-flow: row nowrap;
  }
  .menu{
    flex-flow: row nowrap;
    justify-content: flex-start;
    transform: translateY(0%);
    z-index: initial;
  }

  .menu__link{
    margin: 0 .2rem 0 .2rem;
  }

  .menu__button{
    display: none;
  }

  .menu--show{
    transform: translateY(0%);
  }
}
