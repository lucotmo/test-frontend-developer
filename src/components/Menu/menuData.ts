export const menuData = [
  {
    id: '1',
    name: 'Avianca',
  },
  {
    id: '2',
    name: 'Vivair',
  }
];