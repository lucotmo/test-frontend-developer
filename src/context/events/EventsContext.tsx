import { createContext } from 'react';

interface EventsContextState {
  linkSelect: {
    id: string;
    name: string
  };
  changeLinkSelect: (id: string, name: string) => void
}

const EventsContext = createContext<EventsContextState | null>(null);

export default EventsContext;