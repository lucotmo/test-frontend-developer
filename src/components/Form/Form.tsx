import React, { useContext, useState } from 'react'
import EventsContext from '../../context/events/EventsContext'
import { useForm } from '../../hooks/useForm'
import Input from '../../components/ui/inputs/Input'
import InputSelect from '../../components/ui/inputs/InputSelect'
import Button from '../../components/ui/Button'
import Modal from '../../components/ui/Modal'
import './Form.style.css'

const Form = () => {
  // Manejo del state Modal
  const [showModal, setShowModal] = useState(false)

  // Manejo del state Errors
  const [showErrors, setShowErrors] = useState(false)

  // extraer los valores del context
  const eventsContext = useContext(EventsContext);
  const { linkSelect }: any = eventsContext;

  // valores iniciales del useForm
  const {formulario, handleChange, reset } = useForm({
    Name: '',
    Email: '',
    Cellphone: '',
    Age: ''
  });

  // Destructuración de formulario
  const { Name, Email, Cellphone, Age }: any = formulario;

  // Evento submit del formulario
  const handleSubmit = ( event: React.FormEvent) => {
    event.preventDefault();
    // Errores por defecto
    const errors = { name: false, email: false, cellphone: false, age: false }

    // Validacion del input Name
    if ( /[a-zA-Z]/.test(Name) ) { errors.name = true }
    else { errors.name = false }

    // Validacion del input Email
    if ( /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(Email) ) { errors.email = true } 
    else { errors.email = false }

    // Validacion del input Cellphone
    if ( /[0-9]/.test(Cellphone) ) { errors.cellphone = true }
    else { errors.cellphone = false }

    // Validacion del input Age
    if ( Age !== 'DEFAULT' ) { errors.age = true }
    else { errors.age = false }

    // Validacion de todos lo inputs
    if ( errors.name && errors.email && errors.cellphone && errors.age ) {
      // Ocultando los errores
      setShowErrors(false)
      // Mostrando el Modal
      setShowModal(true)
      // Imprimiendo el formulario
      console.log(formulario)
      // Reseteando el formulario
      reset()
    } else {
      // Mostrando errores
      setShowErrors(true)
    }
  }

  return (
    <>
      <form className="form" onSubmit={handleSubmit}>
        <h2 className="form__title">Bienvenido a <strong>{linkSelect?.name}</strong></h2>
        { showErrors && <small className="form__errors">Ingrese todos los campos</small> }
        <Input
          type="text"
          placeholder="Nombre completo"
          name="Name"
          onChange={handleChange}
          value={Name}
        />
        <Input 
          type="email"
          placeholder="Correo"
          name="Email"
          onChange={handleChange}
          value={Email}
        />
        <Input 
          type="text"
          placeholder="Celular"
          name="Cellphone"
          onChange={handleChange}
          value={Cellphone}
        />
        <InputSelect
          value={Age}
          name="Age"
          onChange={handleChange}
        />
        <Button
          type="submit"
          name="Enviar"
        />
      </form>
      { showModal && <Modal onClick={() => setShowModal(false)} /> }
    </>
  )
}

export default Form
