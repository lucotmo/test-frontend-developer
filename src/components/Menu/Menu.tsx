import { useContext, useState } from 'react';

import EventsContext from '../../context/events/EventsContext';
const {menuData} = require('./menuData')

import './Menu.style.css'

const Menu = () => {
  const [showMenu, setShowMenu] = useState(false)
  
  // Extraer los valores del EventsContext
  const eventsContext = useContext(EventsContext);
  const {Â linkSelect, changeLinkSelect }: any = eventsContext;

  return (
    <nav className="container">
      <ul className={`menu ${showMenu && 'menu--show'}`} >
        {
          menuData.map((el: any) => {
            return <li 
              className={`menu__link ${el.id === linkSelect.id && 'activelink'}`} 
              key={el.id}
              onClick={() => {
                changeLinkSelect( `${el.id}`, `${el.name}` )
                setShowMenu(false)
              }}
            >{el.name}</li>
          })
        }
      </ul>
      <div className="menu__button" onClick={() => setShowMenu( !showMenu )}>
        <div className="menu__line"></div>
        <div className="menu__line"></div>
        <div className="menu__line"></div>
      </div>
    </nav>
  )
}

export default Menu
