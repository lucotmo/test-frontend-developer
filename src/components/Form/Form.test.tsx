import { mount } from 'enzyme';
import Form from './Form';
import { menuData } from '../../components/Menu/menuData';
import EventsContext from '../../context/events/EventsContext';

describe('Pruebas en <Form />', () => {
  const changeLinkSelect = jest.fn();

  const wrapper = mount(
    <EventsContext.Provider value={{
      linkSelect: menuData[0],
      changeLinkSelect
    }}>
      <Form  />
    </EventsContext.Provider>
  )

  test('debe de mostrarse correctamente', () => {
    expect( wrapper ).toMatchSnapshot();
  });
})
