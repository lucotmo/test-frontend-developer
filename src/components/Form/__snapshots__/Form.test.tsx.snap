// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Form /> debe de mostrarse correctamente 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <h2
    className="form__title"
  >
    Bienvenido a 
    <strong>
      Avianca
    </strong>
  </h2>
  <div
    className="content"
  >
    <label
      htmlFor="Name"
    >
      Nombre completo
    </label>
    <input
      className="input"
      id="Name"
      name="Name"
      onChange={[Function]}
      placeholder="Nombre completo"
      type="text"
      value=""
    />
  </div>
  <div
    className="content"
  >
    <label
      htmlFor="Email"
    >
      Correo
    </label>
    <input
      className="input"
      id="Email"
      name="Email"
      onChange={[Function]}
      placeholder="Correo"
      type="email"
      value=""
    />
  </div>
  <div
    className="content"
  >
    <label
      htmlFor="Cellphone"
    >
      Celular
    </label>
    <input
      className="input"
      id="Cellphone"
      name="Cellphone"
      onChange={[Function]}
      placeholder="Celular"
      type="text"
      value=""
    />
  </div>
  <div
    className="content"
  >
    <label
      htmlFor="Age"
    >
      Edad
    </label>
    <select
      className="inputSelect"
      id="Age"
      name="Age"
      onChange={[Function]}
      value=""
    >
      <option
        value="DEFAULT"
      >
        Selecciona Edad
      </option>
      <option
        value="18"
      >
        18
         años
      </option>
      <option
        value="19"
      >
        19
         años
      </option>
      <option
        value="20"
      >
        20
         años
      </option>
      <option
        value="21"
      >
        21
         años
      </option>
      <option
        value="22"
      >
        22
         años
      </option>
      <option
        value="23"
      >
        23
         años
      </option>
      <option
        value="24"
      >
        24
         años
      </option>
      <option
        value="25"
      >
        25
         años
      </option>
      <option
        value="26"
      >
        26
         años
      </option>
      <option
        value="27"
      >
        27
         años
      </option>
      <option
        value="28"
      >
        28
         años
      </option>
      <option
        value="29"
      >
        29
         años
      </option>
      <option
        value="30"
      >
        30
         años
      </option>
      <option
        value="31"
      >
        31
         años
      </option>
      <option
        value="32"
      >
        32
         años
      </option>
      <option
        value="33"
      >
        33
         años
      </option>
      <option
        value="34"
      >
        34
         años
      </option>
      <option
        value="35"
      >
        35
         años
      </option>
      <option
        value="36"
      >
        36
         años
      </option>
      <option
        value="37"
      >
        37
         años
      </option>
      <option
        value="38"
      >
        38
         años
      </option>
      <option
        value="39"
      >
        39
         años
      </option>
      <option
        value="40"
      >
        40
         años
      </option>
      <option
        value="41"
      >
        41
         años
      </option>
      <option
        value="42"
      >
        42
         años
      </option>
      <option
        value="43"
      >
        43
         años
      </option>
      <option
        value="44"
      >
        44
         años
      </option>
      <option
        value="45"
      >
        45
         años
      </option>
      <option
        value="46"
      >
        46
         años
      </option>
      <option
        value="47"
      >
        47
         años
      </option>
      <option
        value="48"
      >
        48
         años
      </option>
      <option
        value="49"
      >
        49
         años
      </option>
      <option
        value="50"
      >
        50
         años
      </option>
      <option
        value="51"
      >
        51
         años
      </option>
      <option
        value="52"
      >
        52
         años
      </option>
      <option
        value="53"
      >
        53
         años
      </option>
      <option
        value="54"
      >
        54
         años
      </option>
      <option
        value="55"
      >
        55
         años
      </option>
      <option
        value="56"
      >
        56
         años
      </option>
      <option
        value="57"
      >
        57
         años
      </option>
      <option
        value="58"
      >
        58
         años
      </option>
      <option
        value="59"
      >
        59
         años
      </option>
      <option
        value="60"
      >
        60
         años
      </option>
      <option
        value="61"
      >
        61
         años
      </option>
      <option
        value="62"
      >
        62
         años
      </option>
      <option
        value="63"
      >
        63
         años
      </option>
      <option
        value="64"
      >
        64
         años
      </option>
      <option
        value="65"
      >
        65
         años
      </option>
      <option
        value="66"
      >
        66
         años
      </option>
      <option
        value="67"
      >
        67
         años
      </option>
      <option
        value="68"
      >
        68
         años
      </option>
      <option
        value="69"
      >
        69
         años
      </option>
      <option
        value="70"
      >
        70
         años
      </option>
      <option
        value="71"
      >
        71
         años
      </option>
      <option
        value="72"
      >
        72
         años
      </option>
      <option
        value="73"
      >
        73
         años
      </option>
      <option
        value="74"
      >
        74
         años
      </option>
      <option
        value="75"
      >
        75
         años
      </option>
      <option
        value="76"
      >
        76
         años
      </option>
      <option
        value="77"
      >
        77
         años
      </option>
      <option
        value="78"
      >
        78
         años
      </option>
      <option
        value="79"
      >
        79
         años
      </option>
      <option
        value="80"
      >
        80
         años
      </option>
      <option
        value="81"
      >
        81
         años
      </option>
      <option
        value="82"
      >
        82
         años
      </option>
      <option
        value="83"
      >
        83
         años
      </option>
      <option
        value="84"
      >
        84
         años
      </option>
      <option
        value="85"
      >
        85
         años
      </option>
      <option
        value="86"
      >
        86
         años
      </option>
      <option
        value="87"
      >
        87
         años
      </option>
      <option
        value="88"
      >
        88
         años
      </option>
      <option
        value="89"
      >
        89
         años
      </option>
      <option
        value="90"
      >
        90
         años
      </option>
      <option
        value="91"
      >
        91
         años
      </option>
      <option
        value="92"
      >
        92
         años
      </option>
      <option
        value="93"
      >
        93
         años
      </option>
      <option
        value="94"
      >
        94
         años
      </option>
      <option
        value="95"
      >
        95
         años
      </option>
      <option
        value="96"
      >
        96
         años
      </option>
      <option
        value="97"
      >
        97
         años
      </option>
      <option
        value="98"
      >
        98
         años
      </option>
      <option
        value="99"
      >
        99
         años
      </option>
      <option
        value="100"
      >
        100
         años
      </option>
    </select>
  </div>
  <button
    className="button"
    type="submit"
  >
    Enviar
  </button>
</form>
`;
